File: scripts/generate_articles.mjs
#!/usr/bin/env node
import { put } from '@vercel/blob';
import { neon } from '@neondatabase/serverless';
import Anthropic from '@anthropic-ai/sdk';
import { readFileSync } from 'fs';

const SITE = 'lawnsguide';
const SCORE_THRESHOLD = 80;
const CONCURRENCY = 3;

const AUTHORS = [
  'james-miller', 'sarah-chen', 'mike-rodriguez', 'emily-watson',
  'david-park', 'lisa-thompson', 'robert-hayes', 'anna-kowalski',
];

const IDEAS = [
  // lawn-care (13)
  ['lawn-care', 'how-to-overseed-a-thin-lawn-in-fall'],
  ['lawn-care', 'best-grass-seed-for-shady-areas'],
  ['lawn-care', 'how-to-dethatch-your-lawn-by-hand-or-machine'],
  ['lawn-care', 'lawn-aeration-when-and-how-to-do-it'],
  ['lawn-care', 'how-to-fix-bare-patches-in-your-lawn'],
  ['lawn-care', 'spring-lawn-care-checklist-for-homeowners'],
  ['lawn-care', 'how-to-edge-a-lawn-for-a-clean-look'],
  ['lawn-care', 'organic-lawn-fertilizer-vs-synthetic-which-is-better'],
  ['lawn-care', 'how-to-water-new-grass-seed-properly'],
  ['lawn-care', 'winter-lawn-care-tips-to-prepare-for-cold'],
  ['lawn-care', 'how-to-level-an-uneven-lawn'],
  ['lawn-care', 'best-lawn-mowers-for-small-yards'],
  ['lawn-care', 'how-to-stripe-your-lawn-like-a-pro'],
  // landscaping (13)
  ['landscaping', 'how-to-design-a-front-yard-landscape-plan'],
  ['landscaping', 'best-ground-cover-plants-for-slopes'],
  ['landscaping', 'how-to-install-landscape-edging-around-flower-beds'],
  ['landscaping', 'rock-garden-ideas-for-low-maintenance-yards'],
  ['landscaping', 'how-to-create-a-rain-garden-in-your-backyard'],
  ['landscaping', 'native-plants-for-landscaping-by-region'],
  ['landscaping', 'how-to-add-curb-appeal-with-simple-landscaping'],
  ['landscaping', 'backyard-fire-pit-area-landscaping-ideas'],
  ['landscaping', 'how-to-landscape-around-a-pool'],
  ['landscaping', 'seasonal-flower-bed-planting-guide'],
  ['landscaping', 'how-to-use-mulch-in-landscaping-correctly'],
  ['landscaping', 'small-backyard-landscaping-ideas-on-a-budget'],
  ['landscaping', 'how-to-create-a-wildlife-friendly-garden'],
  // gardening (14)
  ['gardening', 'how-to-grow-tomatoes-in-containers'],
  ['gardening', 'best-herbs-to-grow-indoors-year-round'],
  ['gardening', 'how-to-grow-strawberries-in-a-small-garden'],
  ['gardening', 'fall-garden-planting-guide-what-to-grow'],
  ['gardening', 'how-to-improve-clay-soil-for-gardening'],
  ['gardening', 'container-gardening-tips-for-beginners'],
  ['gardening', 'how-to-grow-peppers-from-seed-to-harvest'],
  ['gardening', 'best-flowers-to-attract-pollinators-to-your-garden'],
  ['gardening', 'how-to-water-garden-plants-efficiently'],
  ['gardening', 'square-foot-gardening-method-explained'],
  ['gardening', 'how-to-grow-garlic-in-your-backyard'],
  ['gardening', 'winter-gardening-crops-you-can-grow-in-cold-weather'],
  ['gardening', 'how-to-build-a-trellis-for-climbing-plants'],
  ['gardening', 'organic-pest-control-methods-for-vegetable-gardens'],
  // home-garden (13)
  ['home-garden', 'how-to-create-a-cottage-garden-style'],
  ['home-garden', 'best-perennial-flowers-for-low-maintenance-gardens'],
  ['home-garden', 'how-to-design-a-herb-garden-near-your-kitchen'],
  ['home-garden', 'raised-garden-bed-soil-mix-recipe'],
  ['home-garden', 'how-to-grow-roses-for-beginners'],
  ['home-garden', 'backyard-vegetable-garden-layout-ideas'],
  ['home-garden', 'how-to-start-seeds-indoors-before-spring'],
  ['home-garden', 'best-shade-plants-for-north-facing-gardens'],
  ['home-garden', 'how-to-build-a-cold-frame-for-year-round-growing'],
  ['home-garden', 'garden-irrigation-systems-drip-vs-sprinkler'],
  ['home-garden', 'how-to-grow-blueberries-in-your-backyard'],
  ['home-garden', 'vertical-garden-ideas-for-small-spaces'],
  ['home-garden', 'how-to-make-leaf-mold-compost-for-free'],
  // pest-control (14)
  ['pest-control', 'how-to-get-rid-of-aphids-on-garden-plants'],
  ['pest-control', 'natural-ways-to-repel-mosquitoes-in-your-yard'],
  ['pest-control', 'how-to-identify-and-treat-japanese-beetle-damage'],
  ['pest-control', 'best-methods-to-control-slugs-in-the-garden'],
  ['pest-control', 'how-to-keep-deer-out-of-your-garden'],
  ['pest-control', 'how-to-get-rid-of-moles-in-your-lawn'],
  ['pest-control', 'organic-weed-control-methods-that-work'],
  ['pest-control', 'how-to-prevent-and-treat-powdery-mildew'],
  ['pest-control', 'how-to-control-ants-in-garden-beds'],
  ['pest-control', 'best-companion-plants-to-repel-pests-naturally'],
  ['pest-control', 'how-to-deal-with-squirrels-digging-in-your-garden'],
  ['pest-control', 'identifying-common-garden-diseases-and-treatments'],
  ['pest-control', 'how-to-use-neem-oil-for-garden-pest-control'],
  ['pest-control', 'how-to-get-rid-of-crabgrass-permanently'],
  // tree-care (13)
  ['tree-care', 'how-to-care-for-newly-planted-trees'],
  ['tree-care', 'signs-your-tree-needs-to-be-removed'],
  ['tree-care', 'how-to-fertilize-trees-and-shrubs-correctly'],
  ['tree-care', 'best-fast-growing-trees-for-privacy'],
  ['tree-care', 'how-to-treat-tree-diseases-common-problems'],
  ['tree-care', 'how-to-stake-a-young-tree-properly'],
  ['tree-care', 'fruit-tree-pruning-guide-for-home-orchards'],
  ['tree-care', 'how-to-remove-a-tree-stump-yourself'],
  ['tree-care', 'evergreen-trees-for-year-round-privacy-screening'],
  ['tree-care', 'how-to-protect-trees-from-winter-damage'],
  ['tree-care', 'ornamental-trees-for-small-yards'],
  ['tree-care', 'how-to-identify-tree-pests-and-treat-them'],
  ['tree-care', 'when-to-call-an-arborist-vs-diy-tree-care'],
];

const TOPIC_PROMPTS = {
  'lawn-care': 'Lawn maintenance, grass health, mowing, fertilizing, watering. Include specific grass species, product names, application rates, seasonal timing, and references to university extension services or turfgrass research.',
  'landscaping': 'Landscape design, outdoor spaces, plants, hardscaping. Include plant species with Latin names, cost estimates, square footage guidance, and references to landscape architecture principles or ASLA guidelines.',
  'gardening': 'Vegetable and flower gardening, soil, planting, harvesting. Include planting dates by USDA zone, spacing measurements, yield data, and references to university cooperative extension or RHS guidance.',
  'home-garden': 'Home garden design, ornamental plants, garden structures. Include plant hardiness zones, bloom times, soil pH requirements, and references to horticultural societies or botanical gardens.',
  'pest-control': 'Garden and lawn pest management, organic and chemical controls. Include pest lifecycle data, treatment timing, product active ingredients, and references to IPM programs or university entomology departments.',
  'tree-care': 'Tree planting, pruning, health, removal. Include species-specific guidance, growth rates, root spread data, and references to ISA (International Society of Arboriculture) or ANSI A300 standards.',
};

// ── ENV SETUP ─────────────────────────────────────────────────────────────────
const siteEnv = {};
for (const line of readFileSync(`/root/vercel-projects/${SITE}/.env.local`, 'utf8').split('\n')) {
  const m = line.match(/^([^#=]+)=(.*)/);
  if (m) siteEnv[m[1].trim()] = m[2].trim().replace(/^"|"$/g, '');
}
const hermesEnv = {};
for (const line of readFileSync('/root/.hermes/profiles/theme-site-worker/.env', 'utf8').split('\n')) {
  const m = line.match(/^([^#=]+)=(.*)/);
  if (m) hermesEnv[m[1].trim()] = m[2].trim();
}
const sql = neon(siteEnv.DATABASE_URL);
const anthropic = new Anthropic({
  apiKey: hermesEnv.ANTHROPIC_API_KEY,
  baseURL: hermesEnv.ANTHROPIC_BASE_URL || 'https://api.anthropic.com',
});
// ─────────────────────────────────────────────────────────────────────────────

const FORBIDDEN_TITLES = ['About ','Why ','Types and Variants','Key Features','Pros and Cons','How to Choose','Conclusion','FAQs','The Bottom Line','In Summary'];
const FORBIDDEN_PHRASES = ['In conclusion','Comprehensive guide','Ultimate guide','Delve into','Navigating the world','Unveil the secrets',"In today's fast-paced",'Look no further',"Whether you're a beginner",'Dive deep into','Tapestry','Testament to','Embark on a journey'];

function scoreArticle(html) {
  let score = 90;
  const text = html.replace(/<[^>]+>/g, '');
  const headings = [...html.matchAll(/<h[23][^>]*>(.*?)<\/h[23]>/gis)].map(m => m[1].replace(/<[^>]+>/g, ''));
  if (FORBIDDEN_TITLES.some(f => headings.some(h => h.toLowerCase().includes(f.toLowerCase())))) score -= 15;
  if (FORBIDDEN_PHRASES.some(f => html.toLowerCase().includes(f.toLowerCase()))) score -= 15;
  if (text.length < 3000) score -= 10;
  const h2=(html.match(/<h2/g)||[]).length, h3=(html.match(/<h3/g)||[]).length, p=(html.match(/<p/g)||[]).length, ul=(html.match(/<ul|<ol/g)||[]).length, tbl=(html.match(/<table/g)||[]).length, bq=(html.match(/<blockquote/g)||[]).length;
  if (!(h2>=5&&h3>=3&&p>=15&&ul>=2&&(tbl>=1||bq>=1))) score -= 10;
  const nums=(text.match(/\b\d+\.?\d*\s*(?:%|degrees?|inches?|feet|foot|lbs?|pounds?|gallons?|sq\.?\s*ft|mph|psi|weeks?|days?|months?|years?|°[FC]|times?|hours?|minutes?|per\s+\w+)/gi)||[]).length;
  const standalone=(text.match(/\b\d{2,}\b/g)||[]).length;
  if (nums+standalone<5) score -= 10;
  if ((text.match(/(?:University|Institute|USDA|EPA|ISA|Journal|Study|Research|according to)[^.]{0,80}\d{4}/gi)||[]).length<2) score -= 10;
  if ((text.match(/\b[A-Z][a-zA-Z&\s.]+(?:University|College|Institute|Extension|State|County|City|Farm|Garden|Park|Department|Association|Foundation)\b/g)||[]).length<3) score -= 10;
  const sentences=text.split(/[.!?]/).map(s=>s.trim().split(/\s+/).length).filter(l=>l>3);
  if (sentences.length>0){const avg=sentences.reduce((a,b)=>a+b,0)/sentences.length;if(sentences.reduce((a,b)=>a+(b-avg)**2,0)/sentences.length<10)score-=10;}
  return Math.max(0, score);
}

async function generateArticleHtml(type, shortTitle) {
  const title = shortTitle.replace(/-/g,' ').replace(/\b\w/g,c=>c.toUpperCase());
  const topicHint = TOPIC_PROMPTS[type]||'';
  const prompt = `Write a high-quality blog article body.

Title: ${title}
Category: ${type}
Topic guidance: ${topicHint}

REQUIREMENTS:
- Output ONLY the article body — NO title, NO author, NO date, NO byline
- Pure text >= 3000 characters
- 5+ h2, 3+ h3, 15+ p, 2+ ul/ol, 1+ table or blockquote
- 5+ specific data points with numbers/measurements
- 2+ citations with org name and year
- 3+ real named institutions or locations
- Start directly with first h2 or intro paragraph

FORBIDDEN headings: About [X], Why [X] Is Gaining Popularity, Types and Variants, Key Features and Benefits, Pros and Cons, How to Choose, Conclusion, FAQs, The Bottom Line, In Summary
FORBIDDEN phrases: In conclusion, Comprehensive guide, Ultimate guide, Delve into, Navigating the world of, Unveil the secrets, In today's fast-paced, Look no further, Whether you're a beginner, Dive deep into, Tapestry, Testament to, Embark on a journey

Output ONLY the HTML body content, no markdown fences, no wrapper article tags.`;
  const resp = await anthropic.messages.create({ model:'claude-opus-4-6', max_tokens:4096, messages:[{role:'user',content:prompt}] });
  let html = resp.content[0].text.trim();
  html = html.replace(/^```html?\s*/i,'').replace(/\s*```$/,'');
  return html;
}

async function generateCoverImage(shortTitle, type) {
  const fallback = `https://picsum.photos/seed/${shortTitle}/1024/576`;
  const title = shortTitle.replace(/-/g,' ').replace(/\b\w/g,c=>c.toUpperCase());
  const hint = TOPIC_PROMPTS[type]||type;
  const prompt = `Professional blog cover photo for ${title}. ${hint}. Clean editorial style, natural lighting, no text overlay.`;
  try {
    const res = await fetch('https://dashscope.aliyuncs.com/api/v1/services/aigc/multimodal-generation/generation',{
      method:'POST',
      headers:{Authorization:`Bearer ${siteEnv.DASHSCOPE_API_KEY}`,'Content-Type':'application/json'},
      body:JSON.stringify({model:'qwen-image-plus',input:{messages:[{role:'user',content:[{text:prompt}]}]},parameters:{size:'1024*576'}}),
    });
    const data = await res.json();
    const ossUrl = data?.output?.choices?.[0]?.message?.content?.[0]?.image;
    if (!ossUrl) return fallback;
    const imgBuf = Buffer.from(await (await fetch(ossUrl)).arrayBuffer());
    if (imgBuf.length<1024) return fallback;
    const blob = await put(`covers/${SITE}/${shortTitle}.png`,imgBuf,{access:'public',token:siteEnv.BLOB_READ_WRITE_TOKEN,allowOverwrite:true,contentType:'image/png'});
    return blob.url;
  } catch { return fallback; }
}

async function insertArticle(data) {
  await sql`INSERT INTO articles (site,type,short_title,language,published_time,modified_time,author,img,title,description,url,body,tag,is_online) VALUES (${data.site},${data.type},${data.short_title},${data.language},${data.published_time},${data.modified_time},${data.author},${data.img},${data.title},${data.description},${data.url},${data.body},${data.tag},${data.is_online}) ON CONFLICT DO NOTHING`;
}

async function getExisting() {
  const rows = await sql`SELECT short_title FROM articles WHERE site=${SITE}`;
  return new Set(rows.map(r=>r.short_title));
}

async function processOne({idx,total,type,shortTitle,author}) {
  const title = shortTitle.replace(/-/g,' ').replace(/\b\w/g,c=>c.toUpperCase());
  process.stdout.write(`[${idx}/${total}] ${shortTitle} ... `);
  let html=null;
  for (let attempt=0;attempt<3;attempt++) {
    try {
      html = await generateArticleHtml(type,shortTitle);
      const sc = scoreArticle(html);
      process.stdout.write(`score=${sc} `);
      if (sc>=SCORE_THRESHOLD) break;
      process.stdout.write(`(retry ${attempt+1}) `);
      html=null;
    } catch(e) { process.stdout.write(`ERR:${e.message} `); html=null; }
  }
  if (!html) { process.stdout.write('FAIL\n'); return false; }
  process.stdout.write('img... ');
  const img = await generateCoverImage(shortTitle,type);
  process.stdout.write('img OK ');
  const now=new Date(), modified=new Date(now.getTime()+Math.floor(Math.random()*30)*86400000);
  await insertArticle({site:SITE,type,short_title:shortTitle,language:'en',published_time:now,modified_time:modified,author,img,title,description:`Learn about ${title.toLowerCase()} with expert tips and data-backed advice.`,url:`/${type}/${shortTitle}`,body:html,tag:type,is_online:'1'});
  process.stdout.write('DB OK\n');
  return true;
}

async function runWithConcurrency(tasks,limit,fn) {
  const results=[], executing=new Set();
  for (const task of tasks) {
    const p=fn(task).then(r=>{executing.delete(p);return r;});
    executing.add(p); results.push(p);
    if (executing.size>=limit) await Promise.race(executing);
  }
  return Promise.all(results);
}

const existing = await getExisting();
console.log(`Existing in DB: ${existing.size}`);
const tasks = IDEAS.map(([type,shortTitle],i)=>({idx:i+1,total:IDEAS.length,type,shortTitle,author:AUTHORS[i%AUTHORS.length]})).filter(t=>!existing.has(t.shortTitle));
console.log(`To generate: ${tasks.length}`);

let written=0,failed=0;
const results = await runWithConcurrency(tasks,CONCURRENCY,t=>processOne(t));
results.forEach(ok=>ok?written++:failed++);

const rate=written+failed>0?Math.round(written/(written+failed)*100):0;
console.log(`\nArticles Done!\nSite: ${SITE}\nArticles written: ${written}/${IDEAS.length}\nQuality pass rate: ${rate}%\nDB: Neon (ep-fancy-leaf-a4zukau9)\n`);

File: scripts/generate_articles.py
#!/usr/bin/env python3
"""Generate articles for lawnsguide and insert into Neon DB."""

import os, re, time, random, asyncio
from datetime import datetime, timedelta
from dotenv import load_dotenv
import psycopg2

load_dotenv("/root/vercel-projects/lawnsguide/.env.local")
DATABASE_URL = os.environ["DATABASE_URL"]

SITE = "lawnsguide"
START_DATE = datetime(2025, 5, 26)
CONCURRENCY = 5
SCORE_THRESHOLD = 80  # max score is 90 after removing HTML structure check

AUTHORS = [
    "james-miller", "sarah-chen", "mike-rodriguez", "emily-watson",
    "david-park", "lisa-thompson", "robert-hayes", "anna-kowalski"
]

IDEAS = [
    ("lawn-care", "how-often-should-you-water-your-lawn-in-summer"),
    ("lawn-care", "best-time-to-fertilize-lawn-spring-vs-fall"),
    ("lawn-care", "why-your-grass-turns-yellow-and-how-to-fix-it"),
    ("lawn-care", "mowing-height-guide-for-different-grass-types"),
    ("landscaping", "low-maintenance-front-yard-landscaping-ideas"),
    ("landscaping", "how-to-build-a-garden-path-with-stepping-stones"),
    ("landscaping", "backyard-privacy-landscaping-plants-and-fences"),
    ("landscaping", "drought-tolerant-landscaping-ideas-for-dry-climates"),
    ("pest-control", "how-to-get-rid-of-grubs-in-your-lawn-naturally"),
    ("pest-control", "common-lawn-weeds-and-how-to-remove-them"),
    ("pest-control", "signs-of-lawn-fungus-and-treatment-options"),
    ("tree-care", "when-and-how-to-prune-trees-for-healthy-growth"),
    ("tree-care", "best-shade-trees-for-small-backyards"),
    ("tree-care", "how-to-plant-a-tree-correctly-step-by-step"),
    ("gardening", "beginner-vegetable-garden-layout-and-planting-guide"),
    ("gardening", "companion-planting-combinations-that-actually-work"),
    ("gardening", "how-to-start-a-raised-bed-garden-from-scratch"),
    ("home-garden", "essential-lawn-and-garden-tools-every-homeowner-needs"),
    ("home-garden", "outdoor-living-space-ideas-for-small-backyards"),
    ("home-garden", "how-to-compost-at-home-for-a-healthier-garden"),
]

TOPIC_PROMPTS = {
    "lawn-care": "Focus on specific mowing schedules, watering frequencies (gallons/week), fertilizer NPK ratios, and seasonal timing. Include real product names and university extension research citations.",
    "landscaping": "Include specific plant species with Latin names, cost estimates in USD, square footage calculations, and real project examples from specific US regions.",
    "pest-control": "Include pest identification with scientific names, infestation thresholds (insects per sq ft), specific pesticide active ingredients, and integrated pest management (IPM) data.",
    "tree-care": "Include tree species with growth rates (ft/year), pruning timing by USDA hardiness zone, root spread ratios, and arborist certification references (ISA standards).",
    "gardening": "Include planting dates by zone, spacing in inches, yield estimates (lbs per plant), soil pH requirements, and companion planting research citations.",
    "home-garden": "Include tool specifications (blade width, weight), ergonomic ratings, price ranges, maintenance schedules, and real brand comparisons.",
}

FORBIDDEN_TITLES = [
    "About ", "Why ", "Types and Variants", "Key Features", "Pros and Cons",
    "How to Choose", "Conclusion", "FAQs", "The Bottom Line", "In Summary"
]
FORBIDDEN_PHRASES = [
    "In conclusion", "Comprehensive guide", "Ultimate guide", "Delve into",
    "Navigating the world", "Unveil the secrets", "In today's fast-paced",
    "Look no further", "Whether you're a beginner", "Dive deep into",
    "Tapestry", "Testament to", "Embark on a journey"
]

def get_existing(conn):
    with conn.cursor() as cur:
        cur.execute("SELECT short_title FROM articles WHERE site=%s", (SITE,))
        return {r[0] for r in cur.fetchall()}

def score_article(html: str, title: str) -> int:
    score = 90  # max is 90 (HTML structure check removed)
    text = re.sub(r'<[^>]+>', '', html)

    # Forbidden titles (15pts) — only check h2/h3 headings, not full text
    headings = re.findall(r'<h[23][^>]*>(.*?)</h[23]>', html, re.I | re.S)
    heading_text = ' '.join(re.sub(r'<[^>]+>', '', h) for h in headings)
    for f in FORBIDDEN_TITLES:
        if f.lower() in heading_text.lower():
            score -= 15
            break

    # Forbidden phrases (15pts)
    for f in FORBIDDEN_PHRASES:
        if f.lower() in html.lower():
            score -= 15
            break

    # Text length (10pts)
    if len(text) < 3000:
        score -= 10

    # Element counts (10pts)
    h2 = len(re.findall(r'<h2', html))
    h3 = len(re.findall(r'<h3', html))
    p  = len(re.findall(r'<p',  html))
    ul = len(re.findall(r'<ul|<ol', html))
    tbl = len(re.findall(r'<table', html))
    bq  = len(re.findall(r'<blockquote', html))
    if not (h2>=5 and h3>=3 and p>=15 and ul>=2 and (tbl>=1 or bq>=1)):
        score -= 10

    # Data points (10pts) — any number with context
    nums = re.findall(r'\b\d+\.?\d*\s*(?:%|degrees?|inches?|feet|foot|lbs?|pounds?|gallons?|sq\.?\s*ft|mph|psi|weeks?|days?|months?|years?|°[FC]|times?|hours?|minutes?|per\s+\w+)', text, re.I)
    # Also count standalone numbers >= 2 digits as data points
    standalone = re.findall(r'\b\d{2,}\b', text)
    if len(nums) + len(standalone) < 5:
        score -= 10

    # Citations (10pts)
    citations = re.findall(r'(?:University|Institute|USDA|EPA|ISA|Journal|Study|Research|according to)[^.]{0,80}\d{4}', text, re.I)
    if len(citations) < 2:
        score -= 10

    # Real cases (10pts) — named institutions/locations
    named = re.findall(r'\b[A-Z][a-zA-Z&\s\.]+(?:University|College|Institute|Extension|State|County|City|Farm|Garden|Park|Department|Association|Foundation)\b', text)
    if len(named) < 3:
        score -= 10

    # Writing quality (10pts) — basic check: sentence variety
    sentences = re.split(r'[.!?]', text)
    lengths = [len(s.split()) for s in sentences if len(s.split()) > 3]
    if lengths:
        avg = sum(lengths)/len(lengths)
        variance = sum((l-avg)**2 for l in lengths)/len(lengths)
        if variance < 10:
            score -= 10

    return max(0, score)

def load_hermes_env():
    hermes_env = {}
    with open('/root/.hermes/profiles/theme-site-worker/.env') as f:
        for line in f:
            line = line.strip()
            if line and not line.startswith('#') and '=' in line:
                k, v = line.split('=', 1)
                hermes_env[k] = v
    return hermes_env

def generate_article_html(type_: str, short_title: str, title: str) -> str:
    """Call Anthropic claude-opus-4-6 to generate article body HTML (no title/author/date)."""
    import anthropic
    hermes_env = load_hermes_env()
    client = anthropic.Anthropic(
        api_key=hermes_env['ANTHROPIC_API_KEY'],
        base_url=hermes_env.get('ANTHROPIC_BASE_URL', 'https://api.anthropic.com'),
    )

    topic_hint = TOPIC_PROMPTS.get(type_, "")
    prompt = f"""Write a high-quality blog article body for a lawn and garden website.

Title: {title}
Category: {type_}

Topic guidance: {topic_hint}

REQUIREMENTS:
- Output ONLY the article body content — NO title, NO author name, NO date, NO byline
- Pure text >= 3000 characters
- 5+ h2 headings, 3+ h3 headings, 15+ paragraphs
- 2+ ul/ol lists, 1+ table, 1+ blockquote
- 5+ specific data points with numbers/percentages/measurements
- 2+ citations with organization name and year (e.g. "According to USDA research in 2023...")
- 3+ real named locations or institutions (e.g. "Texas A&M University", "Midwest region")
- Start directly with the first h2 or introductory paragraph

FORBIDDEN section headings: About [X], Why [X] Is Gaining Popularity, Types and Variants, Key Features and Benefits, Pros and Cons, How to Choose, Conclusion, FAQs, The Bottom Line, In Summary

FORBIDDEN phrases: "In conclusion", "Comprehensive guide", "Ultimate guide", "Delve into", "Navigating the world of", "Unveil the secrets", "In today's fast-paced", "Look no further", "Whether you're a beginner", "Dive deep into", "Tapestry", "Testament to", "Embark on a journey"

Output ONLY the HTML body content, no markdown fences, no wrapper article tags."""

    resp = client.messages.create(
        model="claude-opus-4-6",
        max_tokens=4096,
        messages=[{"role": "user", "content": prompt}],
        timeout=300,
    )
    html = resp.content[0].text.strip()
    html = re.sub(r'^```html?\s*', '', html)
    html = re.sub(r'\s*```$', '', html)
    return html

def generate_cover_image(short_title: str, type_: str, description: str) -> str:
    """Generate cover image via DashScope, upload to Vercel Blob, return permanent URL."""
    import urllib.request, json, subprocess

    # Read keys from .env.local
    env = {}
    with open(f'/root/vercel-projects/{SITE}/.env.local') as f:
        for line in f:
            line = line.strip()
            if '=' in line and not line.startswith('#'):
                k, v = line.split('=', 1)
                env[k] = v.strip('"')

    api_key = env.get('DASHSCOPE_API_KEY', '')
    blob_token = env.get('BLOB_READ_WRITE_TOKEN', '')

    type_context = TOPIC_PROMPTS.get(type_, 'lawn and garden')
    title_readable = short_title.replace('-', ' ').title()
    prompt = f"Professional blog cover photo for an article about {title_readable}. {type_context}. Clean editorial style, natural lighting, no text overlay."

    data = json.dumps({
        "model": "qwen-image-plus",
        "input": {"messages": [{"role": "user", "content": [{"text": prompt}]}]},
        "parameters": {"size": "1024*576"}
    }).encode()

    req = urllib.request.Request(
        "https://dashscope.aliyuncs.com/api/v1/services/aigc/multimodal-generation/generation",
        data=data,
        headers={"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"},
        method="POST"
    )
    with urllib.request.urlopen(req, timeout=120) as resp:
        result = json.loads(resp.read())
    oss_url = result["output"]["choices"][0]["message"]["content"][0]["image"]

    # Download
    tmp_path = f"/tmp/lawnsguide-{short_title}.png"
    urllib.request.urlretrieve(oss_url, tmp_path)
    if os.path.getsize(tmp_path) < 1024:
        return "https://images.unsplash.com/photo-1501004318641-b39e6451bec6?w=1024&h=576&fit=crop"

    if not blob_token:
        return "https://images.unsplash.com/photo-1501004318641-b39e6451bec6?w=1024&h=576&fit=crop"

    # Upload to Vercel Blob
    result = subprocess.run(
        ["npx", "vercel", "blob", "put", tmp_path,
         "--pathname", f"covers/lawnsguide/{short_title}.png",
         "--access", "public", "--allow-overwrite", "true", "--rw-token", blob_token],
        cwd=f"/root/vercel-projects/{SITE}",
        capture_output=True, text=True, timeout=60
    )
    os.remove(tmp_path)
    if result.returncode != 0:
        return oss_url
    for line in (result.stdout + result.stderr).splitlines():
        if line.startswith("> Success!"):
            return line.split("> Success! ", 1)[1].strip()
    return "https://images.unsplash.com/photo-1501004318641-b39e6451bec6?w=1024&h=576&fit=crop"

def insert_article(data: dict):
    """Open a fresh connection per insert to avoid SSL timeout."""
    conn = psycopg2.connect(DATABASE_URL)
    try:
        with conn.cursor() as cur:
            cur.execute("""
                INSERT INTO articles (site, type, short_title, language, published_time, modified_time,
                    author, img, title, description, url, body, tag, is_online)
                VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)
                ON CONFLICT DO NOTHING
            """, (data['site'], data['type'], data['short_title'], data['language'],
                  data['published_time'], data['modified_time'], data['author'],
                  data['img'], data['title'], data['description'],
                  data['url'], data['body'], data['tag'], data['is_online']))
        conn.commit()
    finally:
        conn.close()

def assign_dates(ideas, start_date):
    """Assign published_time: go backwards from start_date, 2-3 articles/author/day, weekdays only."""
    assignments = []
    date = start_date
    author_idx = 0
    i = 0
    while i < len(ideas):
        if date.weekday() < 5:  # Mon-Fri
            slots = min(3, len(ideas) - i)
            for _ in range(slots):
                # Random time HH:MM:SS during working hours
                pub_dt = date.replace(
                    hour=random.randint(8, 20),
                    minute=random.randint(0, 59),
                    second=random.randint(0, 59)
                )
                assignments.append((ideas[i], AUTHORS[author_idx % len(AUTHORS)], pub_dt))
                author_idx += 1
                i += 1
        date -= timedelta(days=1)
    return assignments

def slug_from_short_title(short_title: str) -> str:
    return short_title.lower().replace(' ', '-')

def title_from_short_title(short_title: str) -> str:
    return short_title.replace('-', ' ').title()

def process_one(args):
    idx, total, type_, short_title, author, pub_date = args
    title = title_from_short_title(short_title)
    print(f"[{idx}/{total}] Generating: {short_title} ...", end=" ", flush=True)

    html = None
    for attempt in range(3):
        try:
            html = generate_article_html(type_, short_title, title)
            sc = score_article(html, title)
            print(f"score={sc}", end=" ", flush=True)
            if sc >= SCORE_THRESHOLD:
                break
            print(f"(retry {attempt+1})", end=" ", flush=True)
            html = None
        except Exception as e:
            print(f"ERROR:{e}", end=" ", flush=True)
            html = None

    if not html:
        print("FAIL", flush=True)
        return False

    description = f"Learn about {title.lower()} with expert tips and data-backed advice."
    img_url = ''
    try:
        print("img...", end=" ", flush=True)
        img_url = generate_cover_image(short_title, type_, description)
        print("img OK", end=" ", flush=True)
    except Exception as e:
        print(f"img FAIL({e})", end=" ", flush=True)

    url = f"/{type_}/{short_title}"
    now = datetime.now()
    mod_date = now + timedelta(days=random.randint(1, 30))
    insert_article({
        'site': SITE, 'type': type_, 'short_title': short_title, 'language': 'en',
        'published_time': now, 'modified_time': mod_date,
        'author': author, 'img': img_url,
        'title': title, 'description': description,
        'url': url, 'body': html, 'tag': type_, 'is_online': '1'
    })
    print(f"DB OK", flush=True)
    return True

def main():
    from concurrent.futures import ThreadPoolExecutor, as_completed

    conn = psycopg2.connect(DATABASE_URL)
    existing = get_existing(conn)
    conn.close()
    print(f"Existing articles in DB: {len(existing)}", flush=True)

    assignments = assign_dates(IDEAS, START_DATE)
    total = len(IDEAS)

    tasks = []
    for idx, ((type_, short_title), author, pub_date) in enumerate(assignments, 1):
        if short_title in existing:
            print(f"[{idx}/{total}] SKIP (exists): {short_title}", flush=True)
            continue
        tasks.append((idx, total, type_, short_title, author, pub_date))

    written = 0
    failed = 0
    with ThreadPoolExecutor(max_workers=3) as executor:
        futures = {executor.submit(process_one, t): t for t in tasks}
        for f in as_completed(futures):
            if f.result():
                written += 1
            else:
                failed += 1

    rate = round((written / (written + failed)) * 100) if (written + failed) > 0 else 0
    print(f"""
Articles Done!

Site: {SITE}
Topics: 6 topics
Articles written: {written}/{total}
Quality pass rate: {rate}%
DB: Neon (ep-fancy-leaf-a4zukau9)
""")

if __name__ == "__main__":
    main()
